lib/AST/Type.cpp

```diff
@@ -3032,7 +3027,8 @@ TypeSubstitutionMap TypeBase::getMemberSubstitutions(const DeclContext *dc) {
   // Find the superclass type with the context matching that of the member.
   //
   // FIXME: Do this in the caller?
-  if (baseTy->getAnyNominal()) {
+  if (!isContextContainedNominalTypeFromSuperClass(baseTy, dc, resolver) &&
```

Comment:
Yeah, skipping this part is not correct in the general case. Imagine if you hit your case above where the dc is actually one of the children of the "correct" dc, but also it is a subclass. You'll end up getting the wrong substitutions for baseTy.